### ISSUE TITLE
Too much Data? More data = Worse Results?

### ISSUE THREAD

**Issue by callmesora** [ultralytics/yolov5#7612]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Can a I have too much Data?

I was training yolov5m with pre-trained weights on the BDD100K Data set.
This Dataset has 100k Images.

![image](https://user-images.githubusercontent.com/61661948/165625973-1c9dc0c4-0b6d-4309-8c46-14c9be0e35f5.png)

They Seem to be a bit **unbalanced in terms of number of instances.**

I've ran the multi-GPU training program and obtained very poor results. Due to the huge ammount of images I could't use the --cache flag. So **each epoc took a huge amout of time.** 

And besides the huge amount of time I also **got terrible results**
I'm not sure why the mAP didn't even reach 1%. And kept getting worse .
**It took 50 Hours to run around 20 Epochs**

![image](https://user-images.githubusercontent.com/61661948/165626250-858a0208-c8bb-4b2d-9add-2caed2c77226.png)

I later tried to use only a sub section of the Dataset with 10K images and I got much better results and managed to cache the images, so training went a lot faster. And I also had better results. I have no Idea why

![image](https://user-images.githubusercontent.com/61661948/165626529-08ecb62f-d42e-4f95-91ab-736a3d0cbca1.png)

### **So In Recap my question(s) are:**

- Why is a bigger dataset performing poorly?
- Why do I have a peak at the epoch number 1000, Did I just leave a local minimum or is it a bug? It has a huge spike **(Red Arrow pointing in the image)**
- Why are my results so poor? mAP of less than 30%
- I Feel like those under represented classes are affecting severly the mAP calculatio**n. Is there a way to only train on specific classes of my dataset?** (Let's say, train on the top 4 most represented)  I don't know a way to remove the classes I don't want from the labels.

**Comment by glenn-jocher** [comment]
@callmesora 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by glenn-jocher** [comment]
@callmesora 👋 Hello! Thanks for asking about **training speed issues**. YOLOv5 🚀 can be trained on CPU (slowest), [single-GPU](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data), or [multi-GPU](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) (fastest). If you would like to increase your training speed some options are:

- Increase `--batch-size`
- Reduce `--img-size`
- Reduce model size, i.e. from YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s
- Train with [multi-GPU](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) DDP at larger `--batch-size`
- Train on cached data: `python train.py --cache` (RAM caching) or `--cache disk` (disk caching)
- Train on faster GPUs, i.e.: P100 -> V100 -> A100
- Train on free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Good luck 🍀 and let us know if you have any other questions!

**Comment by callmesora** [comment]
Hello, I've tried and ajusted all of those previously. I'm traning on multiple GPU as well. (I'm unsure if this is an automated message , I assume so), regardless. 
Thanks for the valuable inputs but I've tried changing those values but my questions still remain, if any of you could help me It would be lovely!

**Comment by yyring000** [comment]
Hi, did you solve this problem, i also got this problem, I use YOLOV5S to detect pedestrains in BDD100K dataset, but the output map, p, r, all are very low. If you can give some advice? thank you!!!

**Comment by MartinPedersenpp** [comment]
What is the format of the BDD100K Data set annotations and how are they converted?
If you want to detect pedestrians, the pretrained Yolov5 models are trained on the coco dataset, which includes plenty of pictures of persons, which might as well be pedestrians.
I hope this helps

**Comment by glenn-jocher** [comment]
@callmesora please submit a PR with a BDD100K.yaml to help other users get started training on this dataset. If we have this yaml and can train on the dataset we can help determine what mAP to expect on the dataset.

Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started.

**Comment by callmesora** [comment]
Hey, I will send you all the scripts and techniques I used to convert and prepare the Dataset. In the meantime I

1. Reduced the size of the Dataset to 10k Images
2. Filter the under represented classes according to Glenn recomendation (10k Instances minimum).
3. I Converged Car Bus Truck and Motorcycle classes into one  class "car" since the last 3 were severely under represented
4. I Dropped some of the labels  such as  "other person" "other vehicle" "train" 
5. I noticed the model was overfitting on Objectness with the default hyper parameters so I dropped it's gain and added more augmentations. 

**All this trippled my mAP (from step 2-5).**
I will provide as soon as possible a **git repo and a PR** with the yaml I used and a **guide to pre process the Data a**nd the scripts I used. 

If someone doesn't want to wait here's how I did it:
Tried to use BDD100K official bdd-to-coco conversion and from coco to YOLOV5 with scripts I found online but No luck, I kept getting some labels messed up or some errors I couldn't solve.

**Then I found another aproach**
I used fifty one library to pre process the BDD100K into the YOLOV5 PyTorch format and used the Validation set since it had 10k Images. 

These 10K images luckly fit the Roboflow Maximum free images per dataset so I used it to merge and drop some of the labels (yes you can drop all but pedestrians if you wish!). 
You can merge and find these when you generate a new "version" of the Dataset in Roboflow, as so:
![image](https://user-images.githubusercontent.com/61661948/166235907-a0c4c4d7-d820-41b2-ba09-a511237fc9fd.png)

I also have a script for dividing the Dataset into train / val folders already in the PyTorch format in case you don't want to be limited by Roboflow 10K images.  This pops a new problem I haven't been able to solve. How to merge and drop classes "manually" I have no script for it and haven't found any library that does it for me. 

**If anyone knows how to drop / merge classes in a YOLOV5 dataset please let me know!**

But I still don't know why it perfomed worse with 100K images :/

**Comment by callmesora** [comment]
Here is the convertion !
@glenn-jocher  how should I structure this notebook in a PR? Should I add te notebook to a specific folder then sumbit the PR ? 
https://github.com/callmesora/BDD100K-to-YOLOV5

**Comment by glenn-jocher** [comment]
@callmesora thanks for the info! We want to generalize as much as possible, so perhaps your work is not suitable here, as you've customized the dataset?

**Comment by callmesora** [comment]
No @glenn-jocher , I provide the original one , just in the YOLOv5 format :) (This script also converts it to COCO and YOLOv4 etc) formats so regardless I think it will be helpfull for people specially to compare different YOLOs to v5 and get better benchmarks

Those later changes are not in the Repo, they are made latter in roboflow if the user wants it. I just left more guidelines to help people manipulate the dataset to their will afterwords.

**Comment by glenn-jocher** [comment]
@callmesora thanks for sharing the notebook! To submit a PR, you could create a new folder such as `datasets/conversion_scripts` within the YOLOv5 repo, and submit your notebook there. Feel free to mention this in the PR description. This would be very helpful for others to compare different YOLOs and to facilitate better benchmarks. Thank you for your valuable contribution!
